website/src/pages/specification/20241027-flagsets.md

```diff
@@ -0,0 +1,182 @@
+- Single retrievers for multiple flag sets _(as proposed in https://github.com/thomaspoignant/go-feature-flag/issues/2314)_.
+
+_Even if out of scope for now, those are interesting options that we may want to implement later._
+
+## Proposed solution
+
+### Solution 1: 1 flag set per file
+
+In this solution we consider that we need at least one file per flag set.  
+All the flags retrieved by a `retriever` _(aka in the same file)_ will be associated to the same flag set.
+
+We can specify the flag set name in the configuration file.  
+As of today, we can still have multiple files for 1 flag set _(by specifying the same `flagSet` name in each file)_,
+and we will have the same mechanism as of today _(with flag overridden in case of flag name collision)_.
+
+#### Flags configuration file
+```yaml
+# config-file1.goff.yaml
+# Syntax used in this example is just for the sake of the example, it is not the final syntax.
+flagSet: flagset-teamA
+
+featureA-enabled:
+  variations:
+      enabled: true
+      disabled: false
+  defaultRule:
+    variation: enabled
+```
+```yaml
+# config-file2.goff.yaml
+flagSet: flagset-teamB
+
+featureA-enabled:
+  variations:
+      enabled: true
+      disabled: false
+  defaultRule:
+    variation: enabled
+```
+
+#### Relay-proxy configuration example
+```yaml
+# ...
+retrievers:
+  - kind: file
+    path: config-file1.goff.yaml
+  - kind: file
+    path: config-file2.goff.yaml
+
+authorizedKeys:
+  evaluation:
+    - apikey1 # owner: userID1
+    - apikey2 # owner: userID2
+  admin:
+    - apikey3
+```
+
+**PRO**
+- It is simple to configure a flag set by putting all the flags at the same place.
+- It is easy to understand which flags are part of a flag set.
+- It is easy to give ownership of a flag set to a team, by giving them access to the file.
+
+**CON**
+- The flag set configuration is at the same place as the flag configuration, so it is possible to edit the flag set name directly inside the file _(risk of overriding another flag set)_.
+- If we want the same flag to be part of multiple flag sets, we need to duplicate the flag in multiple files.
+
+### Solution 2: specify the flag set in the retriever configuration
+
+In this solution we consider that we need at least one file per flag set.  
+All the flags retrieved by a `retriever` _(aka in the same file)_ will be associated to the same flag set.
+
+We associate the flag set to the retriever in the configuration,
+#### Flags configuration file
+```yaml
+# config-file1.goff.yaml
+featureA-enabled:
+  variations:
+      enabled: true
+      disabled: false
+  defaultRule:
+    variation: enabled
+```
+```yaml
+# config-file2.goff.yaml
+featureA-enabled:
+  variations:
+      enabled: true
+      disabled: false
+  defaultRule:
+    variation: enabled
+```
+
+#### Relay-proxy configuration example
+```yaml
+# ...
+retrievers:
+  - kind: file
+    path: config-file1.goff.yaml
+    flagSet: flagset-teamA
```

Comment:
I'm a little confused how this would work in practice though? If tomorrow a new team comes in and they need their own flag set, then GOFF would need to be redeployed again with a new configuration for another retriever. In our companies' deployment/FF solution, we're hoping each development project in each team is an individual flag set to avoid possible conflict, which means every new project on every team that's created would require a production redeployment of go-feature-flag on our side.  Obviously this isn't really maintainable, which is what our [feature request](https://github.com/thomaspoignant/go-feature-flag/issues/2314) aimed to solve.

Disregarding the feature request previously, would another solution such as having the `flagset` key defined per flag work? Each flag could have a `flagset` key that if not provided would be `default` (keeping original functionality), otherwise would set the flag as being owned by that `flagset`. The issue here is flags still have the same name in their source (S3, MongoDB, Redis, etc.), but this doesn't have to be an issue for goff as when it loads flags into its memory it can use the `flagset` to distinguish flags internally.

Reply:
>I'm a little confused how this would work in practice though? If tomorrow a new team comes in and they need their own flag set, then GOFF would need to be redeployed again with a new configuration for another retriever. In our companies' deployment/FF solution, we're hoping each development project in each team is an individual flag set to avoid possible conflict, which means every new project on every team that's created would require a production redeployment of go-feature-flag on our side. Obviously this isn't really maintainable, which is what our https://github.com/thomaspoignant/go-feature-flag/issues/2314 aimed to solve.

@iisharankov your use case is still in my mind but I want to first introduce the notion of flag set and later address your usecase.
All the proposals are still valid with your use case, it will just need to have some kind of meta retriever that will be able to automatically create a flag set.

I place your need for now in the **Out of scope** section, but as soon as we have the flag set ready we will be able to work on this too.

> Disregarding the feature request previously, would another solution such as having the flagset key defined per flag work? Each flag could have a flagset key that if not provided would be default (keeping original functionality), otherwise would set the flag as being owned by that flagset. The issue here is flags still have the same name in their source (S3, MongoDB, Redis, etc.), but this doesn't have to be an issue for goff as when it loads flags into its memory it can use the flagset to distinguish flags internally.

Yes that was my original idea, but we will have limits because of unicity of flag name here.
See the example in my comments here: https://github.com/thomaspoignant/go-feature-flag/pull/2734#issuecomment-2508689545